File: src/cleansia_ios/CleansiaCustomer/Sources/Features/Profile/Data/LanguagePreferenceSync.swift
import CleansiaCore
import Foundation

/// Pushes the display language the user picked onto `User.PreferredLanguageCode`, which is what every
/// server-rendered mail (booking confirmations, reminders, receipts) is written in. Without it the
/// stamp is frozen at whatever signup happened to send.
@MainActor
protocol LanguagePreferenceSync: AnyObject {
    func send(languageCode: String) async

    /// What a beginning session calls. It differs from `send` in the one way that decides whether the
    /// reconcile means anything: it reads the server before comparing. `send` answers from the cached
    /// profile because the picker only runs on a screen that already has one — and that cache is
    /// cleared with the session, so a reconcile trusting it would compare against nothing on the exact
    /// launch it exists for and silently do nothing.
    func reconcile(languageCode: String) async
}

/// `UpdateCurrentUser` is a blind full replace of first name / last name / phone / birth date and only
/// then applies the language, so a language-only push has to replay the rest of the profile verbatim —
/// and must not run at all on a profile that has holes the validators would reject.
enum LanguagePreferencePush {
    static func update(for user: CurrentUserProfile, languageCode: String) -> ProfileUpdate? {
        guard user.isComplete, user.preferredLanguageCode != languageCode else { return nil }
        return ProfileUpdate(
            id: user.id,
            firstName: user.firstName,
            lastName: user.lastName,
            phoneNumber: user.phoneNumber,
            birthDate: user.birthDate,
            languageCode: languageCode
        )
    }
}

@MainActor
final class LiveLanguagePreferenceSync: LanguagePreferenceSync {
    private let repository: UserProfileRepository

    init(repository: UserProfileRepository) {
        self.repository = repository
    }

    /// Silent on failure by design: a display-language tap is not a save the user is waiting on, and the
    /// local switch has already happened. The next profile save re-sends it.
    func send(languageCode: String) async {
        guard let user = repository.currentUser,
              let update = LanguagePreferencePush.update(for: user, languageCode: languageCode)
        else { return }
        _ = await repository.update(update)
    }

    /// A read then a conditional write, and silent for the same reason. A server that already holds this
    /// language costs the read and nothing else — an unconditional write would replay a stale local
    /// profile over the server's on every launch, for a value that was already right. Signed out, the
    /// read is refused by the token store before a request is made.
    func reconcile(languageCode: String) async {
        await repository.refresh()
        await send(languageCode: languageCode)
    }
}

File: src/cleansia_android/customer-app/src/main/java/cz/cleansia/customer/core/settings/LanguagePreferenceSync.kt
package cz.cleansia.customer.core.settings

import cz.cleansia.customer.core.user.CurrentUser
import cz.cleansia.customer.core.user.UserRepository
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Pushes the display language the user picked onto `User.PreferredLanguageCode`, which is what every
 * server-rendered mail (booking confirmations, reminders, receipts) is written in. Without it the
 * stamp is frozen at whatever signup happened to send, and [AppSettingsRepository] alone only ever
 * writes DataStore.
 */
interface LanguagePreferenceSync {
    suspend fun send(languageCode: String)
}

data class LanguagePush(
    val firstName: String,
    val lastName: String,
    val phoneNumber: String,
    val birthDate: String?,
    val languageCode: String,
)

/**
 * `UpdateCurrentUser` still replaces first and last name outright — only phone, birth date, photo and
 * language treat an absent value as "nothing to say" — so a language-only push has to replay the rest
 * of the profile verbatim, and must not run at all on a profile with holes the validators would reject.
 */
object LanguagePreferencePush {
    fun forUser(user: CurrentUser?, languageCode: String): LanguagePush? {
        if (user == null || user.preferredLanguageCode == languageCode) return null
        val phoneNumber = user.phoneNumber?.trim()?.ifBlank { null } ?: return null
        if (user.firstName.isBlank() || user.lastName.isBlank()) return null
        return LanguagePush(
            firstName = user.firstName,
            lastName = user.lastName,
            phoneNumber = phoneNumber,
            birthDate = user.birthDate,
            languageCode = languageCode,
        )
    }
}

@Singleton
class LiveLanguagePreferenceSync @Inject constructor(
    private val userRepository: UserRepository,
) : LanguagePreferenceSync {

    /**
     * Silent on failure by design: a display-language tap is not a save the user is waiting on, the
     * local switch has already happened, and the next profile save re-sends the code anyway.
     */
    override suspend fun send(languageCode: String) {
        val push = LanguagePreferencePush.forUser(userRepository.currentUser.value, languageCode) ?: return
        userRepository.updateCurrentUser(
            firstName = push.firstName,
            lastName = push.lastName,
            phoneNumber = push.phoneNumber,
            birthDate = push.birthDate,
            languageCode = push.languageCode,
        )
    }
}
